File: java/javafind/src/main/java/javafind/FileTypes.java
package javafind;

import org.sqlite.SQLiteConfig;

import java.nio.file.Path;
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.SQLException;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class FileTypes {
    private static final List<FileType> fileTypes = List.of(
            FileType.UNKNOWN,
            FileType.ARCHIVE,
            FileType.AUDIO,
            FileType.BINARY,
            FileType.CODE,
            FileType.FONT,
            FileType.IMAGE,
            FileType.TEXT,
            FileType.VIDEO,
            FileType.XML
    );
    private static final String SELECT_FILE_NAME_ENTRIES = "SELECT name, file_type_id FROM file_name";
    // private static final String SELECT_FILE_TYPE_ID_FOR_FILE_NAME = "SELECT file_type_id FROM file_name WHERE name = ?";
    private static final String SELECT_FILE_TYPE_ID_FOR_EXTENSION = "SELECT file_type_id FROM file_extension WHERE extension = ?";

    private Connection _conn = null;
    private final Map<String, FileType> _extTypeCache;
    private final Map<String, FileType> _nameTypeCache;
    private boolean _nameTypeCacheLoaded = false;

    private Connection getConnection() {
        if (_conn == null) {
            try {
                // This first line is supposedly needed to load the correct driver
                var cls = Class.forName("org.sqlite.JDBC");
                SQLiteConfig config = new SQLiteConfig();
                config.setReadOnly(true);
                _conn = DriverManager.getConnection("jdbc:sqlite:" + FindConfig.XFIND_DB, config.toProperties());
            } catch (SQLException | ClassNotFoundException e) {
                Logger.logError(e.getMessage());
            }
        }
        return _conn;
    }

    public FileTypes() {
        _extTypeCache = new HashMap<>();
        _nameTypeCache = new HashMap<>();
    }

    private Map<String, FileType> getFileTypesForQueryAndParams(final String query, final List<String> params) {
        var conn = getConnection();
        var results = new HashMap<String, FileType>();
        try {
            var stmt = conn.prepareStatement(query);
            for (int i = 0; i < params.size(); i++) {
                stmt.setString(i + 1, params.get(i));
            }
            var rs = stmt.executeQuery();
            while (rs.next()) {
                var key = rs.getString(1);
                int fileTypeId = rs.getInt(2) - 1;
                results.put(key, fileTypes.get(fileTypeId));
            }
        } catch (SQLException e) {
            Logger.logError(e.getMessage());
        }
        return results;
    }

    private void loadNameTypeCache() {
        _nameTypeCache.putAll(getFileTypesForQueryAndParams(SELECT_FILE_NAME_ENTRIES, List.of()));
        _nameTypeCacheLoaded = true;
    }

    private FileType getFileTypeForQueryAndParams(final String query, final List<String> params) {
        var conn = getConnection();
        try {
            var stmt = conn.prepareStatement(query);
            for (int i = 0; i < params.size(); i++) {
                stmt.setString(i + 1, params.get(i));
            }
            var rs = stmt.executeQuery();
            if (rs.next()) {
                int fileTypeId = rs.getInt("file_type_id") - 1;
                return fileTypes.get(fileTypeId);
            }
        } catch (SQLException e) {
            Logger.logError(e.getMessage());
        }
        return FileType.UNKNOWN;
    }

    private FileType getFileTypeForFileName(final String fileName) {
        if (!_nameTypeCacheLoaded) {
            loadNameTypeCache();
        }
        if (_nameTypeCache.containsKey(fileName)) {
            return _nameTypeCache.get(fileName);
        }
//        return getFileTypeForQueryAndParams(SELECT_FILE_TYPE_ID_FOR_FILE_NAME, List.of(fileName));
        return FileType.UNKNOWN;
    }

    private FileType getFileTypeForExtension(final String fileExt) {
        if (fileExt == null || fileExt.isEmpty()) {
            return FileType.UNKNOWN;
        }
        if (_extTypeCache.containsKey(fileExt)) {
            return _extTypeCache.get(fileExt);
        }
        var fileType = getFileTypeForQueryAndParams(SELECT_FILE_TYPE_ID_FOR_EXTENSION, List.of(fileExt));
        _extTypeCache.put(fileExt, fileType);
        return fileType;
    }

    final FileType getFileType(final Path f) {
        var fileType = getFileTypeForFileName(f.getFileName().toString());
        if (fileType != FileType.UNKNOWN) {
            return fileType;
        }
        return getFileTypeForExtension(FileUtil.getExtension(f));
    }

    final boolean isArchiveFile(final Path path) {
        return getFileType(path) == FileType.ARCHIVE;
    }

    final boolean isAudioFile(final Path path) {
        return getFileType(path) == FileType.AUDIO;
    }

    final boolean isBinaryFile(final Path path) {
        return getFileType(path) == FileType.BINARY;
    }

    public final boolean isCodeFile(final Path path) {
        return getFileType(path) == FileType.CODE;
    }

    public final boolean isFontFile(final Path path) {
        return getFileType(path) == FileType.FONT;
    }

    public final boolean isImageFile(final Path path) {
        return getFileType(path) == FileType.IMAGE;
    }

    final boolean isTextFile(final Path path) {
        var fileType = getFileType(path);
        return fileType == FileType.TEXT
                || fileType == FileType.CODE
                || fileType == FileType.XML;
    }

    final boolean isVideoFile(final Path path) {
        return getFileType(path) == FileType.VIDEO;
    }

    final boolean isUnknownFile(final Path path) {
        return getFileType(path) == FileType.UNKNOWN;
    }

    public final boolean isXmlFile(final Path path) {
        return getFileType(path) == FileType.XML;
    }
}

File: scala/scalafind/src/main/scala/scalafind/FileTypes.scala
package scalafind

import org.sqlite.SQLiteConfig
import scalafind.FileType.FileType

import java.io.IOException
import java.nio.file.Path
import scala.collection.mutable
import java.sql.{Connection, DriverManager, SQLException}
import scala.jdk.CollectionConverters.*

object FileType extends Enumeration {
  type FileType = Value
  val Unknown: FileType = Value("unknown")
  val Archive: FileType = Value("archive")
  val Audio: FileType   = Value("audio")
  val Binary: FileType  = Value("binary")
  val Code: FileType    = Value("code")
  val Font: FileType    = Value("font")
  val Image: FileType   = Value("image")
  val Text: FileType    = Value("text")
  val Video: FileType   = Value("video")
  val Xml: FileType     = Value("xml")

  def forName(fileTypeName: String): FileType = {
    try {
      FileType.withName(fileTypeName.trim.toLowerCase)
    } catch {
      case _: NoSuchElementException => Unknown
    }
  }
}

class FileTypes {

  private val fileTypes = List(
    FileType.Unknown, FileType.Archive, FileType.Audio, FileType.Binary, FileType.Code,
    FileType.Font, FileType.Image, FileType.Text, FileType.Video, FileType.Xml
  )

  private val SELECT_FILE_TYPE_ID_FOR_FILE_NAME = "SELECT file_type_id FROM file_name WHERE name = ?"
  private val SELECT_FILE_TYPE_ID_FOR_EXTENSION = "SELECT file_type_id FROM file_extension WHERE extension = ?"

  private var _conn: Option[Connection] = None
  private var _extTypeCache = Map[String, FileType]()
  private var _nameTypeCache = Map[String, FileType]()
  private var _nameTypeCacheLoaded = false

  private def getConnection: Option[Connection] = {
    Class.forName("org.sqlite.JDBC")
    _conn match {
      case Some(c) => _conn
      case None =>
        try {
          val config = SQLiteConfig()
          config.setReadOnly(true)
          _conn = Some(DriverManager.getConnection("jdbc:sqlite:" + FindConfig.XFIND_DB, config.toProperties))
          _conn
        } catch {
          case e: Exception =>
            Common.logError(e.getMessage)
            None
        }
    }
  }

  private def getFileTypesForQueryAndParams(query: String, params: List[String]): Map[String, FileType] = {
    val results = mutable.Map[String, FileType]()
    getConnection match
      case Some(conn) =>
        try {
          val stmt = conn.prepareStatement(query)
          params.zipWithIndex.foreach { case (param, idx) => stmt.setString(idx + 1, param) }
          val rs = stmt.executeQuery()
          while (rs.next()) {
            val key = rs.getString(1)
            val fileTypeId = rs.getInt(2) - 1
            results(key) = fileTypes(fileTypeId)
          }
          Map.empty[String, FileType] ++ results
        } catch {
          case e: SQLException =>
            Common.logError(e.getMessage)
            Map.empty[String, FileType] ++ results
        }
      case None =>
        Map.empty[String, FileType] ++ results
  }

  private def loadNameTypeCache(): Unit = {
    _nameTypeCache = getFileTypesForQueryAndParams("SELECT name, file_type_id FROM file_name", List.empty)
    _nameTypeCacheLoaded = true
  }

  private def getFileTypeForQueryAndParams(query: String, params: List[String]): FileType = {
    getConnection match
      case Some(conn) =>
        try {
          val stmt = conn.prepareStatement(query)
          params.zipWithIndex.foreach { case (param, idx) => stmt.setString(idx + 1, param) }
          val rs = stmt.executeQuery()
          if (rs.next()) {
            val fileTypeId = rs.getInt("file_type_id") - 1
            fileTypes(fileTypeId)
          } else {
            FileType.Unknown
          }
        } catch {
          case e: SQLException =>
            Common.logError(e.getMessage)
            FileType.Unknown
        }
      case None =>
        FileType.Unknown
  }

  private def getFileTypeForFileName(fileName: String): FileType = {
    if (!_nameTypeCacheLoaded) loadNameTypeCache()
    if (_nameTypeCache.contains(fileName)) return _nameTypeCache(fileName)
//    getFileTypeForQueryAndParams(SELECT_FILE_TYPE_ID_FOR_FILE_NAME, List(fileName))
    FileType.Unknown
  }

  private def getFileTypeForExtension(fileExt: String): FileType = {
    if (fileExt.isBlank) return FileType.Unknown
    if (_extTypeCache.contains(fileExt)) return _extTypeCache(fileExt)
    val fileType = getFileTypeForQueryAndParams(SELECT_FILE_TYPE_ID_FOR_EXTENSION, List(fileExt))
    _extTypeCache += (fileExt -> fileType)
    fileType
  }

  def getFileType(path: Path): FileType = {
    val fileTypeForFileName = getFileTypeForFileName(path.getFileName.toString)
    if (fileTypeForFileName != FileType.Unknown) {
      return fileTypeForFileName
    }
    val extension = FileUtil.getExtension(path)
    getFileTypeForExtension(extension)
  }

  def isArchiveFile(path: Path): Boolean = {
    getFileType(path) == FileType.Archive
  }

  def isAudioFile(path: Path): Boolean = {
    getFileType(path) == FileType.Audio
  }

  def isBinaryFile(path: Path): Boolean = {
    getFileType(path) == FileType.Binary
  }

  def isCodeFile(path: Path): Boolean = {
    getFileType(path) == FileType.Code
  }

  def isFontFile(path: Path): Boolean = {
    getFileType(path) == FileType.Font
  }

  def isImageFile(path: Path): Boolean = {
    getFileType(path) == FileType.Image
  }

  def isTextFile(path: Path): Boolean = {
    val fileType = getFileType(path)
    fileType == FileType.Text || fileType == FileType.Code || fileType == FileType.Xml
  }

  def isUnknownFile(path: Path): Boolean = {
    getFileType(path) == FileType.Unknown
  }

  def isVideoFile(path: Path): Boolean = {
    getFileType(path) == FileType.Video
  }

  def isXmlFile(path: Path): Boolean = {
    getFileType(path) == FileType.Xml
  }
}

object FileTypes {
  def isZipArchiveFile(path: Path): Boolean = {
    Set("zip", "jar", "war").contains(FileUtil.getExtension(path))
  }

  def isGzArchiveFile(path: Path): Boolean = {
    Set("gz", "tgz").contains(FileUtil.getExtension(path))
  }

  def isBz2ArchiveFile(path: Path): Boolean = {
    "bz2" == FileUtil.getExtension(path)
  }

  def isTarArchiveFile(path: Path): Boolean = {
    "tar" == FileUtil.getExtension(path)
  }
}
